packages/fiber/src/core/stages.ts

```diff
@@ -0,0 +1,109 @@
+import { MutableRefObject } from 'react'
+import { StoreApi, UseBoundStore } from 'zustand'
+import { RootState } from './store'
+
+export interface UpdateCallback {
+  (state: RootState, delta: number, frame?: THREE.XRFrame | undefined): void
+}
+
+export type UpdateCallbackRef = MutableRefObject<UpdateCallback>
+type Store = UseBoundStore<RootState, StoreApi<RootState>>
+export type UpdateSubscription = { ref: UpdateCallbackRef; store: Store }
+
+export type FixedStageOptions = { fixedStep?: number; maxSubSteps?: number }
+
+export class Stage {
+  name: string
+  subscribers: UpdateSubscription[]
+
+  constructor(name: string) {
+    this.name = name
+    this.subscribers = []
+  }
+
+  frame(delta: number, frame?: THREE.XRFrame | undefined) {
+    const subs = this.subscribers
+
+    for (let i = 0; i < subs.length; i++) {
+      subs[i].ref.current(subs[i].store.getState(), delta, frame)
+    }
+  }
+
+  add(ref: UpdateCallbackRef, store: Store) {
+    this.subscribers = [...this.subscribers, { ref, store }]
+
+    return () => {
+      this.subscribers = this.subscribers.filter((sub) => {
+        return sub.ref !== ref
+      })
+    }
+  }
+}
+
+const FPS_50 = 1 / 50
+
+export class FixedStage extends Stage {
+  private fixedStep: number
+  private maxSubsteps: number
+  private accumulator: number
+  private alpha: number
+
+  constructor(name: string, options?: { fixedStep?: number; maxSubSteps?: number }) {
+    super(name)
+
+    this.fixedStep = options?.fixedStep ?? FPS_50
+    this.maxSubsteps = options?.maxSubSteps ?? 6
+    this.accumulator = 0
+    this.alpha = 0
+  }
+
+  frame(delta: number, frame?: THREE.XRFrame | undefined) {
+    const initialTime = performance.now()
+    let substeps = 0
+
+    this.accumulator += delta
+
+    while (this.accumulator >= this.fixedStep && substeps < this.maxSubsteps) {
+      this.accumulator -= this.fixedStep
+      substeps++
+
+      super.frame(this.fixedStep, frame)
+
+      if (performance.now() - initialTime > this.fixedStep * 200) {
+        // The framerate is not interactive anymore.
+        break
+      }
+    }
+
+    // The accumulator will only be larger than the fixed step if we had to
+    // bail early due to hitting the max substep limit. In that case, we want to
+    // shave off the excess so we don't fall behind next frame.
+    this.accumulator = this.accumulator % this.fixedStep
+    this.alpha = this.accumulator / this.fixedStep
+  }
+
+  set(options: FixedStageOptions) {
+    const { fixedStep, maxSubSteps } = options
+    if (fixedStep) this.fixedStep = fixedStep
+    if (maxSubSteps) this.maxSubsteps = maxSubSteps
+  }
+
+  get() {
+    return {
+      fixedStep: this.fixedStep,
+      maxSubsteps: this.maxSubsteps,
+      accumulator: this.accumulator,
+      alpha: this.alpha,
+    }
+  }
+}
+
+const Early = new Stage('early')
```

Comment:
it seems these strings 'early', 'fixed' etc. are supposed to be used by the devs, i would therfore export an enum to handle this:

```ts
export enum Stage {
  Fixed = 'fixed'
}
```

So they can be used as such `Stage.Fixed`, helps avoid spelling mistakes and a better DX imo.

Reply:
Makes sense, but this name would conflict with the `Stage` class. I'm going to name it `Stages` and also I think rename `pipeline` to `stages` so it is consistent.

Reply:
I'll look into templating for allowing a dev to pass in a custom enum for a custom set of stages as I don't think they can be generated at runtime.

Reply:
Trying to work this out, I'm having trouble with enums and a custom defined lifecycle. The only strategy I can think of is a factory that takes a stages array and enum type generic (as a template) and returns a useUpdate hook. But this would be a big API change though and doesn't seem worth it. Do you have any ideas? @hmans?

Reply:
Is the problem you're trying to solve how to squeeze a `Stage.*` enum into the code that coexists with a `Stage` class? If so, maybe consider just adding a couple of static properties to the `Stage` class, eg.

```
class Stage {
  static Early = 'early'
  static Late = 'late'
  /* ... */
}
```

The functionality is pretty much the same as an enum here. I haven't tried yet, but I think this also allows the user to add extra static names like this to a userland class that extends `Stage` (or not -- I think it's fine if we leave it to the user to decide how to represent their stage names in code.)

Reply:
That's an interesting idea, it works well enough in practice. The challenge was to see if we could have type safety in the stage name parameter for `useUpdate` like Josh suggests below. It would be a nice DX but really seems to overcomplicate things when the user creates their own lifecycle of stages.

Reply:
Oh yeah, I see. I think it would be a good feature to have, but there's probably no way of doing it without either requiring the user to constantly pass a generic type when invoking `useUpdate`, or going through a factory that provides a typed `useUpdate`.

I think it'd be fine to just let this be checked at runtime (ie. throw a warning or error if a `useUpdate` provides a stage name that is not known to the system.) If there's some cool sort of type currying/factory/etc. approach, we can probably still add it later.

Reply:
Sounds good. I sometimes get so sucked into figuring out if I can I forget to ask if I should.

Reply:
Okay, I am exporting a `Stages` enum for the standard lifecycle (I'm trying this word out) and then pushed an example for extending it locally in a custom lifecycle.